### DISCUSSION TITLE
WebBleTransport.listen -> callback doesn't fire

### DISCUSSION THREAD

**Discussion by osabl** [CoolBitX-Technology/coolwallet-js-sdk#279] [Q&A]
I`m trying to use Javascript sdk and connect my coolwallet to my phone using transport-web-ble with instructions from readme. But after pairing a coolwallet to my phone the callback doesn't fire.

My code example.

```
import WebBleTransport from "@coolwallet/transport-web-ble"
import * as core from "@coolwallet/core"

WebBleTransport.listen(async (error, device) => {
    // This callback will never fire 
    const cardName = device.name
    const transport = await WebBleTransport.connect(device)
    const SEPublicKey = await core.config.getSEPublicKey(transport)
    alert(cardName)
    alert(SEPublicKey)
})
```

**Comment by kanitsu** [comment]
Hi osabl.

transport-web-ble, as its name hinted, is used for web browser. It's usually for testing and development purpose, so that you can connect to the wallet using a PC with bluetooth connection and web browser.

If you're building a phone app using react native, you can use transport-react-native-ble instead.

Unfortunately, at the moment the SDK do not support other phone app framework nor native app (app written in java/kotlin/swift).

**Reply by osabl** [reply]
We are building a web app for mobile (PWA) so the app runs in a web browser environment. In theory, this should work on a mobile web browser in the same way as on a desktop. We cannot use transport-react-native-ble because we are not creating a native app, but a web app focused on mobile devices.

**Comment by kanitsu** [comment]
transport-web-ble might not be a suitable for release environment due to its compatibility problem. Here's the list of compatible browsers:
https://developer.mozilla.org/en-US/docs/Web/API/Web_Bluetooth_API#browser_compatibility

I also want to mention that the tutorial on the main page of SDK is outdated. I suggest you refer to transport-web-ble packages instead:
https://github.com/CoolBitX-Technology/coolwallet-sdk/tree/master/packages/transport-web-ble

**Reply by osabl** [reply]
What mean `appId` and where can I get it?

**Comment by kanitsu** [comment]
appId is returned by the card after registration (Please refer to SDK readme on how to register the card to your App).
[Register application with CoolWallet Pro](https://github.com/CoolBitX-Technology/coolwallet-sdk#register-application-with-coolwallet-pro)

**Reply by osabl** [reply]
Ok, but for register card in my app I need has `SEPublicKey` right?
And for getting `SEPublicKey` I need use this code, right?
```
WebBleTransport.listen(async (error, device) => {
    // This callback will never fire
    const transport = await WebBleTransport.connect(device)
    const SEPublicKey = await core.config.getSEPublicKey(transport)
    alert(SEPublicKey)
})
```

But this callback will never fire

**Reply by osabl** [reply]
How I can get `SEPublicKey`?

**Reply by osabl** [reply]
And what is `apdu`?

**Reply by kanitsu** [reply]
`SEPublicKey` along with `appId` is given by the card after successful connection, i.e. the callback fire.
The callback not called is strongly indicates that you bumped into compatibility problem.
As mentioned before, WebBLE is designed for web browser on PC, for development purposes.
I suggest you to run examples in coin-tester on a PC first, to get a grasp on how the card and APP interact.

**Reply by osabl** [reply]
Sorry, but this callback don`t fire even on desktop. 
Here is my code exapmle
```
async testCoolWallet () {
    WebBleTransport.listen(async (error, device) => {
      const cardName = device.name
      const transport = await WebBleTransport.connect(device)
      const SEPublicKey = await core.config.getSEPublicKey(transport)
      console.log(cardName, transport, SEPublicKey)
    })
  },
```
When I run this method I see this
<img width="346" alt="image" src="https://user-images.githubusercontent.com/47601147/170677329-dcfecf2b-1b64-47da-bbed-42408b60cb11.png">
And after I click _Connect_ nothing happens

**Reply by kanitsu** [reply]
What OS you're using? It is known that MacOS have bluetooth problem (if I remember correctly started with Big Sur), which is fixed in Monterey 12.3.1.
